src/core/core.cpp

```diff
@@ -362,49 +363,64 @@ bool Core::checkConnection()
     return toxConnected;
 }
 
+void Core::bootstrapTo(const DhtServer& dhtServer, const QHostAddress& address)
+{
+    ASSERT_CORE_THREAD;
+
+    const auto& ip = address.toString().toUtf8();
+    const auto& pk = dhtServer.publicKey;
+
+    qDebug() << "Connecting to bootstrap node" << pk.toString() << "at" << ip;
+
+    Tox_Err_Bootstrap error;
+    if (dhtServer.statusUdp) {
+        tox_bootstrap(tox.get(), ip.constData(), dhtServer.udpPort, pk.getData(), &error);
+        PARSE_ERR(error);
+    }
+    if (dhtServer.statusTcp) {
+        const auto ports = dhtServer.tcpPorts.size();
+        const auto tcpPort = dhtServer.tcpPorts[rand() % ports];
+        tox_add_tcp_relay(tox.get(), ip.constData(), tcpPort, pk.getData(), &error);
+        PARSE_ERR(error);
+    }
+}
+
 /**
  * @brief Connects us to the Tox network
  */
 void Core::bootstrapDht()
 {
     ASSERT_CORE_THREAD;
 
-
     const auto shuffledBootstrapNodes =
         shuffleBootstrapNodes(bootstrapListGenerator.getBootstrapNodes());
     if (shuffledBootstrapNodes.empty()) {
         qWarning() << "No bootstrap node list";
         return;
     }
 
-    // i think the more we bootstrap, the more we jitter because the more we overwrite nodes
-    auto numNewNodes = 2;
+    // I think the more we bootstrap, the more we jitter because the more we overwrite nodes.
+    int numNewNodes = 2;
     for (int i = 0; i < numNewNodes && i < shuffledBootstrapNodes.size(); ++i) {
         const auto& dhtServer = shuffledBootstrapNodes.at(i);
-        QByteArray address;
-        if (!dhtServer.ipv4.isEmpty()) {
-            address = dhtServer.ipv4.toLatin1();
-        } else if (!dhtServer.ipv6.isEmpty() && settings.getEnableIPv6()) {
-            address = dhtServer.ipv6.toLatin1();
+        QString address;
+        if (!dhtServer.ipv6.isEmpty() && settings.getEnableIPv6()) {
```

Comment:
I think we should coin flip if both ipv4 and ipv6 are available.

Since the list is shuffled already, `i % 2 == 0` should do it

Reply:
I was thinking to prefer ipv6 if people choose it, but maybe we should just use both of them always? If ipv6 is disabled, then only ipv4, otherwise both?

Reply:
how does toxcore handle adding a bs node with a preexisting key?